### DISCUSSION TITLE
Using the Infinispan Client - remove() does not return value from cache

### DISCUSSION THREAD

**Discussion by schall1337** [quarkusio/quarkus#41034] [Q&A] [answered]
Hello,
I am trying out the infinispan client from https://github.com/quarkusio/quarkus-quickstarts.git -> infinispan-client-quickstart

I added an additional endpoint **'deleteGreeting'** to delete entries from cache in the class **InfinispanGreetingResource**
`@Path("/greeting")
public class InfinispanGreetingResource {

    @Inject
    @Remote("mycache")
    RemoteCache<String, Greeting> cache;

    @POST
    @Path("/{id}")
    public CompletionStage<String> postGreeting(String id, Greeting greeting) {
        return cache.putAsync(id, greeting)
              .thenApply(g -> "Greeting added!")
              .exceptionally(ex -> ex.getMessage());
    }

    @GET
    @Path("/{id}")
    public CompletionStage<Greeting> getGreeting(String id) {
        return cache.getAsync(id);
    }

    @DELETE
    @Path("/{id}")
    public String deleteGreeting(String id) {
        Greeting removed = cache.remove(id);
        if(removed != null){
            return "DELETED Successfully!";
        }
        return "No object found to be deleted!";
    }
}`

When call the method **postGreeting** I can see that an entry was successfully added to the cache with the message 'Greeting added!' and in the infinispan console (GUI).

Now when I call the endpoint **deleteGreeting** cache.remove(id); returns null and I see the message "No object found to be deleted!"; 
Though in the infinispan console (GUI) the entry was removed from the cache.

The doc states https://docs.jboss.org/infinispan/15.0/apidocs/org/infinispan/commons/api/BasicCache.html#remove(java.lang.Object)
Specified by:
[remove](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Map.html#remove(java.lang.Object)) in interface [Map](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Map.html)<[K](https://docs.jboss.org/infinispan/15.0/apidocs/org/infinispan/commons/api/BasicCache.html),[V](https://docs.jboss.org/infinispan/15.0/apidocs/org/infinispan/commons/api/BasicCache.html)>
Returns:
the previous value associated with key, or null if there was no mapping for key

What am I doing wrong? Why is cache.remove(id); returning null and not the removed value for given key?

[a comment by an automated account is not shown]

**Reply by karesti** [reply, marked as answer]
@schall1337 try 
```java
cache.withFlags(Flag.FORCE_RETURN_VALUE).remove(id)
```

**Reply by karesti** [reply]
The explanation for this: bringing the removed value has a cost (in production env), so by default the value is not returned. however if you need the deleted value, you can use the flag, so the removed value is returned.

**Comment by schall1337** [comment]
makes sense, somehow I overlooked that options! Thanks very much for the answer and your work.
